OpInfos for stft, istft, fftshift, ifftshift

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68199
* __->__ #68198

This unearths some bugs in istft backward, so I've disabled
backward tests but it's fixed in the next PR in the stack.

cc @mruberry @peterbell10

Differential Revision: [D32467044](https://our.internmc.facebook.com/intern/diff/D32467044)

peterbell10 committed:
OpInfos for stft, istft, fftshift, ifftshift

This unearths some bugs in istft backward, so I've disabled
backward tests but it's fixed in the next PR in the stack.

[ghstack-poisoned]

mruberry commented:
@pytorchbot ciflow rerun -l ciflow/all -l ciflow/slow

mruberry commented:
Failures on ROCm from the next PR in the stack look real:

```
04:43:26 ======================================================================
04:43:26 ERROR [0.198s]: test_fn_grad_istft_cuda_complex128 (__main__.TestGradientsCUDA)
04:43:26 ----------------------------------------------------------------------
04:43:26 Traceback (most recent call last):
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1447, in wrapper
04:43:26     method(*args, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1447, in wrapper
04:43:26     method(*args, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
04:43:26     raise rte
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
04:43:26     result = test(self, **param_kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 772, in dep_fn
04:43:26     return fn(slf, *args, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
04:43:26     return test(*args, **kwargs)
04:43:26   File "test_ops.py", line 664, in test_fn_grad
04:43:26     self._grad_test_helper(device, dtype, op, op.get_op())
04:43:26   File "test_ops.py", line 649, in _grad_test_helper
04:43:26     check_batched_grad=check_batched_grad)
04:43:26   File "test_ops.py", line 627, in _check_helper
04:43:26     check_undefined_grad=check_undefined_grad))
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 2784, in gradcheck
04:43:26     return torch.autograd.gradcheck(fn, inputs, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1329, in gradcheck
04:43:26     return _gradcheck_helper(**args)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1345, in _gradcheck_helper
04:43:26     check_backward_ad=check_backward_ad, nondet_tol=nondet_tol)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1001, in _gradcheck_real_imag
04:43:26     rtol, atol, check_grad_dtypes, nondet_tol)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1222, in _fast_gradcheck
04:43:26     inputs, outputs, func, all_v, all_u, rtol, atol, test_imag, is_forward_ad=use_forward_ad)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1201, in _check_analytical_numerical_equal
04:43:26     raise GradcheckError(_get_notallclose_msg(a, n, j, i, complex_indices, test_imag, is_forward_ad) + jacobians_str)
04:43:26 torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
04:43:26 numerical:tensor(0.0423-0.0025j, device='cuda:0')
04:43:26 analytical:tensor(0.0423-0.0344j, device='cuda:0')
04:43:26 
04:43:26 The above quantities relating the numerical and analytical jacobians are computed 
04:43:26 in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
04:43:26 about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:
04:43:26 
04:43:26 Numerical:
04:43:26  tensor([[ 0.0500+0.0500j,  0.0500-0.0500j],
04:43:26         [ 0.0500+0.0500j,  0.0500-0.0500j],
04:43:26         [-0.1000+0.0000j, -0.0809+0.0588j],
04:43:26         [-0.0309-0.0951j, -0.0809-0.0588j],
04:43:26         [ 0.1000+0.0000j,  0.0309-0.0951j],
04:43:26         [-0.0809+0.0588j,  0.0309+0.0951j],
04:43:26         [-0.1000+0.0000j,  0.0309+0.0951j],
04:43:26         [ 0.0809+0.0588j,  0.0309-0.0951j],
04:43:26         [ 0.1000+0.0000j, -0.0809-0.0588j],
04:43:26         [ 0.0309-0.0951j, -0.0809+0.0588j],
04:43:26         [-0.0500+0.0500j,  0.0500+0.0500j],
04:43:26         [-0.0500+0.0500j,  0.0500+0.0500j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j],
04:43:26         [ 0.0000+0.0000j,  0.0000+0.0000j]], device='cuda:0')
04:43:26 Analytical:
04:43:26 tensor([[ 0.0500+0.0000e+00j,  0.0500+0.0000e+00j],
04:43:26         [ 0.0500+0.0000e+00j,  0.0500+0.0000e+00j],
04:43:26         [-0.1000+1.2660e-18j, -0.0809+5.8779e-02j],
04:43:26         [-0.0309-9.5106e-02j, -0.0809-5.8779e-02j],
04:43:26         [ 0.1000+9.2144e-19j,  0.0309-9.5106e-02j],
04:43:26         [-0.0809+5.8779e-02j,  0.0309+9.5106e-02j],
04:43:26         [-0.1000+9.2144e-19j,  0.0309+9.5106e-02j],
04:43:26         [ 0.0809+5.8779e-02j,  0.0309-9.5106e-02j],
04:43:26         [ 0.1000+1.2660e-18j, -0.0809-5.8779e-02j],
04:43:26         [ 0.0309-9.5106e-02j, -0.0809+5.8779e-02j],
04:43:26         [-0.0500+0.0000e+00j,  0.0500+0.0000e+00j],
04:43:26         [-0.0500+0.0000e+00j,  0.0500+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j],
04:43:26         [ 0.0000+0.0000e+00j,  0.0000+0.0000e+00j]], device='cuda:0')
04:43:26 
04:43:26 The max per-element difference (slow mode) is: 0.050000000000050004.
04:43:26 
04:43:26 
04:43:26 ======================================================================
04:43:26 ERROR [0.182s]: test_fn_grad_istft_cuda_float64 (__main__.TestGradientsCUDA)
04:43:26 ----------------------------------------------------------------------
04:43:26 Traceback (most recent call last):
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1447, in wrapper
04:43:26     method(*args, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1447, in wrapper
04:43:26     method(*args, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
04:43:26     raise rte
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 371, in instantiated_test
04:43:26     result = test(self, **param_kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 772, in dep_fn
04:43:26     return fn(slf, *args, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 737, in test_wrapper
04:43:26     return test(*args, **kwargs)
04:43:26   File "test_ops.py", line 664, in test_fn_grad
04:43:26     self._grad_test_helper(device, dtype, op, op.get_op())
04:43:26   File "test_ops.py", line 649, in _grad_test_helper
04:43:26     check_batched_grad=check_batched_grad)
04:43:26   File "test_ops.py", line 627, in _check_helper
04:43:26     check_undefined_grad=check_undefined_grad))
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 2784, in gradcheck
04:43:26     return torch.autograd.gradcheck(fn, inputs, **kwargs)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1329, in gradcheck
04:43:26     return _gradcheck_helper(**args)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1345, in _gradcheck_helper
04:43:26     check_backward_ad=check_backward_ad, nondet_tol=nondet_tol)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1001, in _gradcheck_real_imag
04:43:26     rtol, atol, check_grad_dtypes, nondet_tol)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1222, in _fast_gradcheck
04:43:26     inputs, outputs, func, all_v, all_u, rtol, atol, test_imag, is_forward_ad=use_forward_ad)
04:43:26   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 1201, in _check_analytical_numerical_equal
04:43:26     raise GradcheckError(_get_notallclose_msg(a, n, j, i, complex_indices, test_imag, is_forward_ad) + jacobians_str)
04:43:26 torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
04:43:26 numerical:tensor(0.0271, device='cuda:0')
04:43:26 analytical:tensor(-0.0126, device='cuda:0')
04:43:26 
04:43:26 The above quantities relating the numerical and analytical jacobians are computed 
04:43:26 in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
04:43:26 about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:
04:43:26 
04:43:26 Numerical:
04:43:26  tensor([[ 0.0500,  0.0500],
04:43:26         [ 0.0500, -0.0500],
04:43:26         [ 0.0500,  0.0500],
04:43:26         [ 0.0500, -0.0500],
04:43:26         [-0.1000, -0.0809],
04:43:26         [ 0.0000,  0.0588],
04:43:26         [-0.0309, -0.0809],
04:43:26         [-0.0951, -0.0588],
04:43:26         [ 0.1000,  0.0309],
04:43:26         [ 0.0000, -0.0951],
04:43:26         [-0.0809,  0.0309],
04:43:26         [ 0.0588,  0.0951],
04:43:26         [-0.1000,  0.0309],
04:43:26         [ 0.0000,  0.0951],
04:43:26         [ 0.0809,  0.0309],
04:43:26         [ 0.0588, -0.0951],
04:43:26         [ 0.1000, -0.0809],
04:43:26         [ 0.0000, -0.0588],
04:43:26         [ 0.0309, -0.0809],
04:43:26         [-0.0951,  0.0588],
04:43:26         [-0.0500,  0.0500],
04:43:26         [ 0.0500,  0.0500],
04:43:26         [-0.0500,  0.0500],
04:43:26         [ 0.0500,  0.0500],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000],
04:43:26         [ 0.0000,  0.0000]], device='cuda:0')
04:43:26 Analytical:
04:43:26 tensor([[ 5.0000e-02,  5.0000e-02],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 5.0000e-02,  5.0000e-02],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [-1.0000e-01, -8.0902e-02],
04:43:26         [ 1.2660e-18,  5.8779e-02],
04:43:26         [-3.0902e-02, -8.0902e-02],
04:43:26         [-9.5106e-02, -5.8779e-02],
04:43:26         [ 1.0000e-01,  3.0902e-02],
04:43:26         [ 9.2144e-19, -9.5106e-02],
04:43:26         [-8.0902e-02,  3.0902e-02],
04:43:26         [ 5.8779e-02,  9.5106e-02],
04:43:26         [-1.0000e-01,  3.0902e-02],
04:43:26         [ 9.2144e-19,  9.5106e-02],
04:43:26         [ 8.0902e-02,  3.0902e-02],
04:43:26         [ 5.8779e-02, -9.5106e-02],
04:43:26         [ 1.0000e-01, -8.0902e-02],
04:43:26         [ 1.2660e-18, -5.8779e-02],
04:43:26         [ 3.0902e-02, -8.0902e-02],
04:43:26         [-9.5106e-02,  5.8779e-02],
04:43:26         [-5.0000e-02,  5.0000e-02],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [-5.0000e-02,  5.0000e-02],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00],
04:43:26         [ 0.0000e+00,  0.0000e+00]], device='cuda:0')
04:43:26 
04:43:26 The max per-element difference (slow mode) is: 0.04999999999999449.
04:43:26 
04:43:26 
04:43:27 ----------------------------------------------------------------------
04:43:27 Ran 10086 tests in 3108.737s
```

That's interesting. Let's add a skip for these tests on ROCm and link an issue in the comment.

mruberry reviewed (approved):
Looks great but see errors; a skip + an issue should let us merge this PR

peterbell10 committed:
Update on "OpInfos for stft, istft, fftshift, ifftshift"


This unearths some bugs in istft backward, so I've disabled
backward tests but it's fixed in the next PR in the stack.

cc mruberry peterbell10

[ghstack-poisoned]

peterbell10 commented:
@mruberry test skip has worked and issue opened in #68429

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8503,6 +8559,45 @@ def ref_pairwise_distance(input1, input2):
+                            'TestJit', 'test_variant_consistency_jit'),
+           ],
+           dtypes=floating_and_complex_types(),
+           # FIXME: col2im does not support automatic differentiation for outputs with complex dtype.
```
And the next PR in the stack doesn't resolve this?

peterbell10 replied:
It does. That's why I remove the fixme note in the next pr.

mruberry replied:
Right; thanks; I just missed that

mruberry commented:
Cool @peterbell10; just one question here and a quick ask to put the URL into the skip comment for istft ROCm autograd

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32467044).
